static/modules/collision.js

```diff
@@ -73,195 +68,554 @@ export function collideWithWorld(entity) {
+  /** @type {Box[]} */
+  const hitTerrain = [];
 
   // Iterate through each colliding entity, and get a vector that defines how
   // "collided" they are
-  for (let i = 0; i < collidingEntities.length; i++) {
-    const collisionVector = calculateCollisionVector(
-      entity,
-      collidingEntities[i]
+  for (let i = 0; i < terrainCollision.length; i++) {
+    const collisionVector = collide(
+      entity.getTerrainCollisionShape(),
+      terrainCollision[i]
     );
+
+    // A block shouldn't be breakable if it is surrounded, though it should
+    // still be collidable for diagonal edge cases.
+    const isBreakable =
+      terrainCollision[i].collidesTop ||
+      terrainCollision[i].collidesBottom ||
+      terrainCollision[i].collidesRight ||
+      terrainCollision[i].collidesLeft;
+
     if (!collisionVector.isZeroVec()) {
       collisionVectors.push(collisionVector);
-      hitEntities.push(collidingEntities[i]);
+      if (isBreakable) {
+        hitTerrain.push(terrainCollision[i]);
+      }
     }
   }
   if (entity.occludedByWalls) {
     // Keep track of how far entity moved during adjustment.
-    const mv = new Vector(0, 0);
+    let mv = new Vector(0, 0);
+
+    /** TODO: this is used to prevent multiple of the same collision from
+     * different wall tiles from being applied. The "correct" solution is to
+     * build the "world" collision objects better but that's difficult and slow.
+     */
+    let alreadyAppliedVectorsX = [];
+    let alreadyAppliedVectorsY = [];
+
     // For each colliding vector, resolve the collision.
     for (let i = 0; i < collisionVectors.length; i++) {
       const cv = collisionVectors[i];
-      if (Math.abs(cv.x) > Math.abs(cv.y) && cv.x != 0) {
-        // If x is the "easiest" solution (but not 0), use x.
-        if (Math.abs(cv.x) > Math.abs(mv.x)) {
-          mv.x = cv.x;
-        }
-      } else if (Math.abs(cv.y) > Math.abs(cv.x) && cv.y != 0) {
-        // If y is the "easiest" solution (but not 0), use y.
-        if (Math.abs(cv.y) > Math.abs(mv.y)) {
-          mv.y = cv.y;
-        }
-      } else {
-        // If X and Y are equal, resolve them both.
-        entity.pos.x -= cv.x;
-        entity.pos.y -= cv.y;
+      if (!alreadyAppliedVectorsX.includes(cv.x)) {
+        mv.x += cv.x;
+        alreadyAppliedVectorsX.push(cv.x);
+      }
+      if (!alreadyAppliedVectorsY.includes(cv.y)) {
+        mv.y += cv.y;
+        alreadyAppliedVectorsY.push(cv.y);
       }
     }
 
     entity.pos = entity.pos.sub(mv);
 
     // bounce based on the move vector
     if (entity.reflectsOffWalls) {
-      if (mv.x !== 0) entity.vel.x *= -1;
-      if (mv.y !== 0) entity.vel.y *= -1;
-      if (hitEntities.length > 0 && entity.wallReflectSpeed !== 0) {
-        entity.vel = entity.vel.norm2().mult(entity.wallReflectSpeed);
+      if (hitTerrain.length > 0 && entity.wallReflectSpeed !== 0) {
+        entity.vel = entity.vel.norm2().mult(-entity.wallReflectSpeed);
```

Comment:
If an entity has `reflectsOffWalls = true` but `wallReflectSpeed = 0` it should reflect with the same speed it hit the wall with. See the description of `wallReflectSpeed` in `entity.js`